proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+By relying on borrowing, `Span` can provide simultaneous access to a non-copyable container, and can help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see the future directions for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+`Span` is the currency type for local processing over values in contiguous memory. It is the replacement for any API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that does not need to escape the value.
+
+### `ContiguousStorage`
+
+A type can indicate that it can provide a `Span` by conforming to the `ContiguousStorage` protocol. `ContiguousStorage` forms a bridge between multi-type or generically-typed interfaces and a performant concrete implementation.
+
+For example, for the hypothetical base64 decoding library mentioned above, a possible API could be:
+
+```swift
+extension HypotheticalBase64Decoder {
+  public func decode(bytes: some ContiguousStorage<UInt8>) -> [UInt8]
+}
+```
+
+Even better, an interface can be defined in terms of the concrete type `Span<UInt8>`:
+
+```swift
+extension Hypothetical Base64Decoder {
+  public func decode(bytes: Span<UInt8>) -> [UInt8]
+}
+```
+
+Advanced libraries might add use an inlinable generic-dispatch interface in addition to a concrete interface defined in terms of `Span`
+
+### `RawSpan`
+
+`RawSpan` allows sharing the contiguous internal representation for values which may be heterogenously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straight-forwards understanding of performance for library code.
+
+`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property or a constructor.
+
+## Detailed design
+
+`Span<Element>` is a simple representation of a span of initialized memory.
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+  public var indices: Range<Int> { get }
```

Comment:
I think this is well-motivated by the `for` loop, but I will note that this might preclude us from formally defining an `Index` type in the future depending on how non-escapable collections go. We could also defer this.

Reply:
We can give it a different name, such as `offsets`, but it feels like using `indices` for _this_ type isn't particularly risky.

Reply:
I'll remove `indices` from the proposal. In the prototype it's now `_indices` and `_byteOffsets` for `Span` and `RawSpan` respectively.